src/coffea/dataset_tools/filespec.py

```diff
@@ -0,0 +1,441 @@
+                    self.root[k] = CoffeaROOTFileSpec(v)
+                if type(v) in [CoffeaParquetFileSpecOptional]:
+                    self.root[k] = CoffeaParquetFileSpec(v)
+                preprocessed[k] = True
+            except Exception:
+                pass
+        if all(preprocessed.values()):
+            try:
+                self.root = PreprocessedFiles(self.root).root
+            except Exception:
+                pass
+        return self
+
+
+class PreprocessedFiles(
+    RootModel[
+        dict[
+            str,
+            Union[
+                CoffeaROOTFileSpec,
+                CoffeaParquetFileSpec,
+            ],
+        ]
+    ],
+    DictMethodsMixin,
+    InputFilesMixin,
+):
+
+    def __iter__(self) -> Iterable[str]:
+        return iter(self.root)
+
+    @model_validator(mode="after")
+    def promote_and_check_files(self) -> Self:
+        preprocessed = {k: False for k in self.root}
+        for k, v in self.root.items():
+            try:
+                if type(v) in [CoffeaROOTFileSpecOptional]:
+                    self.root[k] = CoffeaROOTFileSpec(v)
+                if type(v) in [CoffeaParquetFileSpecOptional]:
+                    self.root[k] = CoffeaParquetFileSpec(v)
+                preprocessed[k] = True
+            except Exception:
+                pass
+        if all(preprocessed.values()):
+            return self
+
+
+class DatasetSpec(BaseModel):
+    files: InputFiles
+    metadata: dict[Hashable, Any] | None = None
+    format: str | None = None
+    form: str | None = None
+
+    @model_validator(mode="before")
+    def preprocess_data(cls, data: Any) -> Any:
+        data = copy.deepcopy(data)
+        if isinstance(data, dict):
+            files = data.pop("files")
+            # promote files list to dict if necessary
+            if isinstance(files, list):
+                # If files is a list, convert it to a dict and let it pass through the rest of the promotion logic
+                tmp = [f.rsplit(":", maxsplit=1) for f in files]
+                files = {}
+                for fsplit in tmp:
+                    # Need a valid split into file name and object path
+                    if len(fsplit) > 1:
+                        # but ensure we don't catch 'root://' and split that
+                        if fsplit[1].startswith("//"):
+                            # no object path
+                            files[":".join(fsplit)] = None
+                        else:
+                            # file name and object path
+                            files[fsplit[0]] = fsplit[1]
+            data["files"] = files
+        elif isinstance(data, DatasetSpec):
+            data = data.model_dump()
+        elif not isinstance(data, DatasetSpec):
+            raise ValueError(
+                "DatasetSpec expects a dictionary with a 'files' key or a DatasetSpec instance"
+            )
+        return data
+
+    def check_form(self) -> bool | None:
```

Comment:
Seems like this could be a validator that maybe also caches the decompressed and deserialized form for later access.

edit: ok later I see this is called by a validator. It should be private then, I think